compiler-rt/lib/builtins/fp_lib.h

```diff
@@ -368,10 +368,11 @@ static __inline fp_t __compiler_rt_fmax(fp_t x, fp_t y) {
 #endif
 }
 
-#elif defined(QUAD_PRECISION) && defined(CRT_HAS_TF_MODE)
+#elif defined(QUAD_PRECISION)
+#if defined(CRT_HAS_TF_MODE)
+#if defined(CRT_HAS_IEEE_TF)
```

Comment:
```suggestion
#if defined(CRT_HAS_IEEE_TF) && defined(CRT_HAS_INT128)
```
Could we simplify this to just be 
```
#if defined(CRT_HAS_IEEE_TF) && defined(CRT_HAS_INT128)
...
#elif defined(CRT_LDBL_128BIT)
...
#else
#error Unsupported TF mode type  
#endif
```

Reply:
I don't think that #if structure you suggest works with sparc.  The logic of the sequence isn't semantically accurate either because the error in the else clause says all of the other conditions above are various TF mode and CRT_LDBL_128BIT is not one.

We will also be losing the check for CRT_HAS_TF_MODE.  

The logic I tried to capture in my changes is "if has a TF mode pick the mode (or give an error), otherwise if has 128-bit long double ...".  I don't think we can flatten this.

Reply:
CRT_HAS_TF_MODE is defined as follows (f128 and int128):
```
// __(u)int128_t is currently needed to compile the *tf builtins as we would
// otherwise need to manually expand the bit manipulation on two 64-bit value.
#if defined(CRT_HAS_128BIT) && defined(CRT_HAS_F128)
#define CRT_HAS_TF_MODE
#endif
```
CRT_HAS_F128 has two options: CRT_HAS_IEEE_TF or CRT_LDBL_128BIT so I believe these checks are equivalent.

We only need the int128 support for the compiler-rt internal IEEE128 functions, for the non-IEEE we fall back to libc calls.

So in my proposed structure, I believe this would work for 32-bit sparc: it has 128-bit float which is ieee but no int128. But since it also has 128-bit long double it would take the second branch and use the libc functions that don't depend on int128.

Reply:
I tried out your proposed structure, and it works for z/OS.  I'll make the change.  Shall I change this:
```
#else
#error Unsupported TF mode type
```
The condition isn't entirely correct anymore since the `CRT_LDBL_128BIT` case wasn't a TF mode.  The if/elif chain handles more than just the different TF modes.

If so, any suggestions?

Reply:
Thanks for testing! I guess we could do something more generic like "TF mode cannot be supported in current configuration", but keeping the current error seems fine to me.